### ISSUE TITLE
Seq2SeqTransformer inputs?

### ISSUE THREAD

**Issue by antony66** [tensorflow/models#10761]
Hey guys,

I am stuck with [Seq2SeqTransformer model](https://github.com/tensorflow/models/blob/master/official/nlp/modeling/models/seq2seq_transformer.py) training.

First of all, typical seq2seq model training requires 2 inputs: encoder sequences and decoder input sequences with start and end tags as well as decoder output sequences without start tag. However, your model expects "inputs" and "targets" as input and also outputs or targets to train the model. What's the difference between input "targets" and output targets? If I throw to the model encoder input sequences, decoder input sequences with starts and ends and output sequences without starts, I only get about 50% accuracy and a trash output.

My training code looks like this:

```
def preprocess_batch(inputs, targets):
    inputs = tf.cast(tokenizer(inputs), tf.int32)
    targets = tf.cast(tokenizer(targets), tf.int32)
    return ({
            "inputs": inputs, 
            "targets": targets[:, :-1]
        }, 
        targets[:, 1:]
    )

dataset = tf.data.Dataset.from_tensor_slices((df['name'], df['normalized']))

split_pos = int(len(dataset) * 0.9)

train_ds = dataset.take(split_pos).batch(BATCH_SIZE)
train_ds = train_ds.map(preprocess_batch, num_parallel_calls=tf.data.AUTOTUNE)
train_ds = train_ds.shuffle(BATCH_SIZE*10).prefetch(16).cache()

val_ds = dataset.skip(split_pos).take(len(dataset)-split_pos).batch(BATCH_SIZE)
val_ds = val_ds.map(preprocess_batch, num_parallel_calls=tf.data.AUTOTUNE)
val_ds = val_ds.shuffle(BATCH_SIZE*10).prefetch(16).cache()

vocab = tokenizer.get_vocabulary()
vocab_size = len(vocab)

start_tag_id = tokenizer(config.start_tag)[0].numpy()
end_tag_id = tokenizer(config.end_tag)[0].numpy()
start_tag_id, end_tag_id

transformer = nlp.models.Seq2SeqTransformer(
    vocab_size=vocab_size,
    embedding_width=config.hidden_dim,
    padded_decode=True,
    decode_max_length=config.decoder_maxlen,
    encoder_layer=nlp.models.TransformerEncoder(
        num_layers=1,
        num_attention_heads=config.num_heads,
        intermediate_size=config.dff,
        use_bias=True,
        norm_first=False,
    ),
    decoder_layer=nlp.models.TransformerDecoder(
        num_layers=4,
        num_attention_heads=config.num_heads,
        intermediate_size=config.dff,
        dropout_rate=0.5,
        use_bias=True,
        norm_first=False,
    ),
    eos_id=end_tag_id,
    name="transformer",
)

learning_rate = CustomSchedule(config.hidden_dim)
optimizer = tf.keras.optimizers.Adam(learning_rate, beta_1=0.9, beta_2=0.98, epsilon=1e-9)

transformer.compile(
    optimizer=optimizer, loss="sparse_categorical_crossentropy", metrics=["accuracy"]
)

early_stopping = tf.keras.callbacks.EarlyStopping(patience=2, restore_best_weights=True)
transformer.fit(train_ds, epochs=10, validation_data=val_ds, callbacks=[early_stopping])
```
And here is the result:

```
Epoch 1/10
3629/3629 [==============================] - 2241s 616ms/step - loss: 3.6709 - accuracy: 0.3267 - val_loss: 4.8040 - val_accuracy: 0.0040
Epoch 2/10
3629/3629 [==============================] - 2187s 603ms/step - loss: 4.8040 - accuracy: 0.0110 - val_loss: 4.8040 - val_accuracy: 0.0040
Epoch 3/10
3629/3629 [==============================] - 2198s 606ms/step - loss: 4.8040 - accuracy: 0.0110 - val_loss: 4.8040 - val_accuracy: 0.0040
```

What am I doing wrong?

My setup: tf: 2.8.0 (apple metal version), tfm installed from pip.

**Comment by gcuder** [comment]
The problem might be in `preprocess_batch` where you shift the `decoder_inputs` to the right. This is done inside the model at this [position](https://github.com/tensorflow/models/blob/7d14c7cac38ad010a6aff41571988dc649ad9ed5/official/nlp/modeling/models/seq2seq_transformer.py#L264).  

The function should be more like this: 

```python
def preprocess_batch(inputs, targets):
    inputs = tf.cast(tokenizer(inputs), tf.int32)
    targets = tf.cast(tokenizer(targets), tf.int32)
    return ({
            "inputs": inputs, 
            "targets": targets
        }, 
        targets
    )

```

Could you try this please and report, if your results get better?

**Comment by antony66** [comment]
@gcuder unfortunately, the results are not any better. This version of `preprocess_batch` is where I actually started exploring this model. During the 1st epoch of training the model reaches accuracy 0.5 but then it looks like the backpropagation just stops and loss starts growing back as well as accuracy starts decreasing. In the end it looks like this:

```
Epoch 1/10
3629/3629 [==============================] - 2440s 671ms/step - loss: 4.3288 - sparse_categorical_accuracy: 0.3597 - val_loss: 4.8040 - val_sparse_categorical_accuracy: 0.0076
Epoch 2/10
 179/3629 [>.............................] - ETA: 35:34 - loss: 4.8040 - sparse_categorical_accuracy: 0.0070
```

However, my goal is to compare performance and accuracy of various seq2seq models built on various libraries so the same dataset works perfectly with KerasNLP based transformer as well as a custom transformer from [this article](https://www.tensorflow.org/text/tutorials/transformer). The latter gives me the following results after 1st epoch:

    Epoch 1 Loss 0.5698, Accuracy 0.8509, time taken 557.3472990989685 secs

Also notice the training time: it's less than 10 minutes per epoch while tfm Seq2SeqTransformer takes about 35 minutes per epoch. KerasNLP transformer is even slower though. I'm guessing this is because of the positional encoding layer in both tfm and KerasNLP models being Embedding layer while the custom one uses sin/cos encoding so it has way less training weights.

**Comment by gcuder** [comment]
That is indeed very interesting. Could you please share a small, self-contained example code where I can reproduce this issue?

**Comment by antony66** [comment]
Sure. This is [the dataset](https://drive.google.com/file/d/1MEZTeFz3YSAn7t2L-KuQRpZAShu5x5gP/view?usp=sharing). It contains 516k rows of randomly generated fake names so no personal data here. And this is [the notebook](https://drive.google.com/file/d/1wqfJGsojdnN1eNUmmBkBIB7kLnN33Vdw/view?usp=sharing). 

I can't reproduce the issue with loss growing back. However, the results are still trash.

**Comment by antony66** [comment]
This is the [Keras NLP version](https://drive.google.com/file/d/1HQ67cFWV8ZTkrgJTbO0HiXh7JZ-pl2ji/view?usp=sharing) just for comparison.

**Comment by gcuder** [comment]
Thank you, could you please make a Google Colab notebook?

**Comment by antony66** [comment]
These are Google colab notebooks. Just click to open them in Colab at the top.

**Comment by antony66** [comment]
Ohh damn it! I have found where my mistake is. This model outputs are logits. So the model needs to be wrapped in order to use `sparse_categorical_crossentropy` loss:

```
transformer_main = nlp.models.Seq2SeqTransformer(
    vocab_size=vocab_size,
    embedding_width=hidden_dim,
    padded_decode=True,
    decode_max_length=decoder_maxlen,
    encoder_layer=nlp.models.TransformerEncoder(
        num_layers=1,
        num_attention_heads=num_heads,
        intermediate_size=dff,
        use_bias=True,
        norm_first=False,
    ),
    decoder_layer=nlp.models.TransformerDecoder(
        num_layers=4,
        num_attention_heads=num_heads,
        intermediate_size=dff,
        dropout_rate=0.5,
        use_bias=True,
        norm_first=False,
    ),
    eos_id=end_tag_id,
    name="transformer",
)

encoder_inputs = tf.keras.Input(shape=(None,), dtype=tf.int32, name="inputs")
decoder_inputs = tf.keras.Input(shape=(None,), dtype=tf.int32, name="targets")
outputs = transformer_main(({'inputs': encoder_inputs, 'targets': decoder_inputs}))
outputs = tf.keras.layers.Dense(vocab_size, activation="softmax")(outputs)
transformer = tf.keras.Model([encoder_inputs, decoder_inputs], outputs, name="transformer")
```

I am sorry. This is my bad.
